src/Microsoft.IdentityModel.Tokens/Json/JsonSerializerPrimitives.cs

```diff
@@ -976,93 +976,97 @@ public static void WriteObject(ref Utf8JsonWriter writer, string key, object obj
 
             Type objType = obj.GetType();
 
-            if (obj is string str)
-                writer.WriteString(key, str);
-            else if (obj is long l)
-                writer.WriteNumber(key, l);
-            else if (obj is int i)
-                writer.WriteNumber(key, i);
-            else if (obj is bool b)
-                writer.WriteBoolean(key, b);
-            else if (obj is DateTime dt)
-                writer.WriteString(key, dt.ToUniversalTime().ToString("O", CultureInfo.InvariantCulture));
-            else if (typeof(IDictionary).IsAssignableFrom(objType))
+            switch (obj)
             {
-                IDictionary dictionary = (IDictionary)obj;
-                writer.WritePropertyName(key);
-
-                writer.WriteStartObject();
-                foreach (var k in dictionary.Keys)
-                    WriteObject(ref writer, k.ToString(), dictionary[k]);
-
-                writer.WriteEndObject();
-            }
-            else if (typeof(IList).IsAssignableFrom(objType))
-            {
-                IList list = (IList)obj;
-                writer.WriteStartArray(key);
-                foreach (var k in list)
-                    WriteObjectValue(ref writer, k);
-
-                writer.WriteEndArray();
-            }
-            else if (obj is JsonElement j)
-            {
-                writer.WritePropertyName(key);
-                j.WriteTo(writer);
-            }
-            else if (obj is double dub)
-                // Below net6.0, we have to convert the double to a decimal otherwise values like 1.11 will be serailized as 1.1100000000000001
-                // large and small values such as double.MaxValue and double.MinValue cannot be converted to decimal.
-                // In these cases, we will write the double as is.
+                case string str:
+                    writer.WriteString(key, str);
+                    break;
+                case long l:
+                    writer.WriteNumber(key, l);
+                    break;
+                case int i:
+                    writer.WriteNumber(key, i);
+                    break;
+                case bool b:
+                    writer.WriteBoolean(key, b);
+                    break;
+                case DateTime dt:
+                    writer.WriteString(key, dt.ToUniversalTime().ToString("O", CultureInfo.InvariantCulture));
+                    break;
+                case IDictionary dictionary:
+                    writer.WritePropertyName(key);
+                    writer.WriteStartObject();
+                    foreach (var k in dictionary.Keys)
+                        WriteObject(ref writer, k.ToString(), dictionary[k]);
+                    writer.WriteEndObject();
+                    break;
+                case IList list:
```

Comment:
Is this the same as IsAssignableFrom?

Reply:
Good point, probably not quite the same, since I'm no longer making changes to this part of the code (the reason I changed it in the first place), I'll just revert it.